packages/calypso-e2e/src/lib/pages/full-site-editor-page.ts

```diff
@@ -0,0 +1,170 @@
+import { Locator, Page } from 'playwright';
+import {
+	BlockInserter,
+	EditorGutenbergComponent,
+	EditorSidebarBlockInserterComponent,
+	EditorToolbarComponent,
+	EditorWelcomeTourComponent,
+} from '..';
+import { getCalypsoURL } from '../../data-helper';
+import envVariables from '../../env-variables';
+
+const wpAdminPath = 'wp-admin/themes.php';
+
+const selectors = {
+	editorIframe: `iframe.is-loaded[src*="${ wpAdminPath }"]`,
+	editorRoot: 'body.block-editor-page',
+	editorCanvasIframe: 'iframe[name="editor-canvas"]',
+	editorCanvasRoot: '.wp-site-blocks',
+	templateLoadingSpinner: '[aria-label="Block: Template Part"] .components-spinner',
+};
+
+/**
+ * Represents an instance of the FSE site editor.
+ * This class is composed of editor components, combining them into larger flows.
+ */
+export class FullSiteEditorPage {
+	private page: Page;
+
+	private editorToolbarComponent: EditorToolbarComponent;
+	private editorGutenbergComponent: EditorGutenbergComponent;
+	private editorSidebarBlockInserterComponent: EditorSidebarBlockInserterComponent;
+	private editorWelcomeTourComponent: EditorWelcomeTourComponent;
+
+	/**
+	 * Constructs an instance of the page POM class.
+	 *
+	 * @param {Page} page The underlying page.
+	 */
+	constructor( page: Page ) {
+		this.page = page;
+
+		this.editorToolbarComponent = new EditorToolbarComponent( page, this.editor );
+		this.editorWelcomeTourComponent = new EditorWelcomeTourComponent( page, this.editor );
+		this.editorSidebarBlockInserterComponent = new EditorSidebarBlockInserterComponent(
+			page,
+			this.editor
+		);
+		// Because of the unique extra iframe in the site editor, this component needs the canvas locator.
+		this.editorGutenbergComponent = new EditorGutenbergComponent( page, this.editorCanvas );
+	}
+
+	/**
+	 * A frame-safe locator to the top level element in the editor, for building other editor locators.
+	 */
+	private get editor(): Locator {
+		return this.page.url().includes( wpAdminPath )
+			? this.page.locator( selectors.editorRoot )
+			: this.page.frameLocator( selectors.editorIframe ).locator( selectors.editorRoot );
+	}
+
+	/**
+	 * A frame-safe locator to the top level element in the editor canvas, for building other editor locators.
+	 * The editor canvas is an iframe wrapping the center editor area that is unique to the site-editor.
+	 */
+	private get editorCanvas(): Locator {
+		return this.editor
+			.frameLocator( selectors.editorCanvasIframe )
+			.locator( selectors.editorCanvasRoot );
+	}
+
+	/**
+	 * Visit the site editor by URL directly.
+	 *
+	 * @param siteHostName Host name of the site, without scheme. (e.g. testsite.wordpress.com)
```

Comment:
I haven't tested yet, but it's not entirely clear to me how the `constructor` is able to get hold of editor components in the DOM before the `visit` method is called (i.e before the editor page is actually loaded), but I'm probably missing something 🤔

Reply:
>how the constructor is able to get hold of editor components in the DOM before the visit method is called (i.e before the editor page is actually loaded)

Ah yes! So, with the new `Locator` [API](https://playwright.dev/docs/locators), Playwright fundamentally changed the mental model behind finding elements. 

Essentially, Locators are _instructions_ on how to find elements, rather than pointers to DOM elements themselves. That allows you to define them at any point, and effectively delay their ultimate DOM querying until later.

The local `editor` properties are essentially just instructions on how to find the top layer of the editor. Because they're just instructions, we can define them whenever during construction, and pass those instructions down to subcomponents!

Definitely a weird mental model shift (it was for me at least), but I think in the end it makes for some much easier patterns!

Reply:
Ah, understood! Thanks for elaborating!